OpenRA.Game/UtilityCommands/ExtractChromeStrings.cs

```diff
@@ -0,0 +1,320 @@
+		{
+			var nodeSplit = node.Key.Split('@');
+			var nodeType = nodeSplit[0];
+			var nodeId = nodeSplit.Length > 1 ? ClearContainersAndToLower(nodeSplit[1]) : null;
+
+			if ((nodeType == "Background" || nodeType == "Container") && nodeId != null)
+				container = nodeId;
+
+			// Get translatable types.
+			var validChildTypes = new List<(MiniYamlNodeBuilder Node, string Type, string Value)>();
+			foreach (var childNode in node.Value.Nodes)
+			{
+				if (translatables.ContainsKey(nodeType))
+				{
+					var childType = childNode.Key.Split('@')[0];
+					if (translatables[nodeType].Contains(childType)
+						&& !string.IsNullOrEmpty(childNode.Value.Value)
+						&& !IsAlreadyTranslated(childNode.Value.Value)
+						&& childNode.Value.Value.Any(char.IsLetterOrDigit))
+					{
+						var translationValue = childNode.Value.Value
+							.Replace("\\n", "\n    ")
+							.Replace("{", "<")
+							.Replace("}", ">")
+							.Trim().Trim('\n');
+
+						validChildTypes.Add((childNode, childType.ToLowerInvariant(), translationValue));
+					}
+				}
+			}
+
+			// Generate translation key.
+			if (validChildTypes.Count > 0)
+			{
+				nodeType = ClearTypesAndToLower(nodeType);
+
+				var translationKey = nodeType;
+				if (!string.IsNullOrEmpty(container))
+				{
+					var containerType = string.Join('-', container.Split('_').Exclude(nodeType).Where(s => !string.IsNullOrEmpty(s)));
+					if (!string.IsNullOrEmpty(containerType))
+						translationKey = $"{translationKey}-{containerType}";
+				}
+
+				if (!string.IsNullOrEmpty(nodeId))
+				{
+					nodeId = string.Join('-', nodeId.Split('_')
+						.Except(string.IsNullOrEmpty(container) ? new string[] { nodeType } : container.Split('_').Append(nodeType))
+						.Where(s => !string.IsNullOrEmpty(s)));
+
+					if (!string.IsNullOrEmpty(nodeId))
+						translationKey = $"{translationKey}-{nodeId}";
+				}
+
+				foreach (var (childNode, childType, translationValue) in validChildTypes)
+					translations.Add(new TranslationCandidate(translationKey, childType, translationValue.Trim().Trim('\n'), childNode));
+			}
+
+			// Recurse.
+			foreach (var childNode in node.Value.Nodes)
+				if (childNode.Key == "Children")
+					foreach (var n in childNode.Value.Nodes)
+						FromChromeLayout(n, translatables, container, ref translations);
+		}
+
+		/// <summary>This is a helper method to find untranslated strings in chrome layouts.</summary>
+		public static void FindUntranslatedStringFields(ModData modData)
+		{
+			var types = modData.ObjectCreator.GetTypes();
+			foreach (var (type, fields) in types.Where(t => t.Name.EndsWith("Widget", StringComparison.InvariantCulture) && t.IsSubclassOf(typeof(Widget))).ToDictionary(t => t.Name[..^6],
+				t => t.GetFields().Where(f => f.Name != "Id" && f.IsPublic && f.FieldType == typeof(string) && !f.HasAttribute<TranslationReferenceAttribute>()).Distinct().Select(f => f.Name).ToList()))
+				if (fields.Count > 0)
```

Comment:
This is also soup. Line breaks are free, throw some in.

Same query r.e. Distinct.